scripts/alert_emails/send_emails.ts

```diff
@@ -162,73 +168,95 @@ async function setLastSnapshotNumber(
-                const currentData = (
-                  await db.collection('alerts-subscriptions').doc(doc.id).get()
-                ).data();
-                if (currentData) {
-                  await db
-                    .collection('invalid-alerts-subscriptions')
-                    .doc(doc.id)
-                    .set(currentData);
-                  await db
-                    .collection('alerts-subscriptions')
-                    .doc(doc.id)
-                    .delete();
-                }
-              } else {
-                errorCount += 1;
-              }
-            });
-        }
-      });
+      .get();
+    return querySnapshot.docs.map(emailDoc => emailDoc.id);
+  };
+
+  async function onInvalidEmail(email: string) {
+    invalidEmailCount += 1;
+    const querySnapshot = await db
+      .collection('alerts-subscriptions')
+      .doc(email)
+      .get();
+    const currentData = querySnapshot.data();
+    if (currentData) {
+      await db
+        .collection('invalid-alerts-subscriptions')
+        .doc(email)
+        .set(currentData);
+      await db.collection('alerts-subscriptions').doc(email).delete();
+    }
+  }
+
+  async function onEmailSent(email: string, fips: string) {
+    emailSent += 1;
+    uniqueEmailAddress[email] = (uniqueEmailAddress[email] || 0) + 1;
+    locationsWithEmails[fips] = (locationsWithEmails[fips] || 0) + 1;
+    // comment out this code if you are developing locally
+    return db
+      .collection(`snapshots/${currentSnapshot}/locations/${fips}/emails/`)
+      .doc(email)
+      .set({ sentAt: admin.firestore.FieldValue.serverTimestamp() });
+  }
+
+  async function sendAlertEmail(email: string, fips: string) {
+    const locationAlert = locationsWithAlerts[fips];
+    const sendData = generateSendData(email, locationAlert);
+
+    try {
+      const response = await sendEmail(api, sendData, dryRun);
+      await onEmailSent(email, fips);
+      return response;
+    } catch (err) {
+      if (isInvalidEmail(err)) {
+        await onInvalidEmail(email);
+      } else {
+        errorCount += 1;
+      }
+    }
   }
+
+  async function sendEmailBatch(emails: string[], fips: string) {
+    if (emails.length === 0) {
+      return;
+    }
+
+    const [firstEmail, ...remainingEmails] = emails;
+    const pilotReq = await sendAlertEmail(firstEmail, fips);
+
+    const rateLimitRemaining = parseInt(pilotReq.headers['x-ratelimit-limit']);
+    const rateLimitResetSec =
+      parseInt(pilotReq.headers['x-ratelimit-reset']) + 15;
+
+    if (rateLimitRemaining < remainingEmails.length + 10) {
+      console.log(
+        `Rate limit remaining: ${rateLimitRemaining}. Sleeping ${rateLimitResetSec} seconds.`,
+      );
+      await delay(rateLimitResetSec * 1000);
+    }
+
+    await Promise.all(
+      remainingEmails.map(email => {
+        return sendAlertEmail(email, fips);
+      }),
+    );
+  }
+
+  for (const fips of locations) {
```

Comment:
I think I mentioned this before but it would be more optimal if we precomputed a list of `[email, fips]` tuples and then we called `_.chunk()` on that so we get full parallelization, even across fips. Perhaps add a TODO for now.